src/App.js

```diff
@@ -26,40 +27,109 @@ const generateSquares = () => {
 };
 
 const App = () => {
-  // This starts state off as a 2D array of JS objects with
-  // empty value and unique ids.
   const [squares, setSquares] = useState(generateSquares());
-
+  const [ticDict, setTicDict] = useState({});
+  const [turn, setTurn] = useState(PLAYER_1);
+  const [winner, setWinner] = useState(null);
   // Wave 2
   // You will need to create a method to change the square
   //   When it is clicked on.
   //   Then pass it into the squares as a callback
 
+  const toggleTurn = () => {
+    let newTurn = '';
+    if (turn === PLAYER_1) {
+      newTurn = PLAYER_2;
+    } else {
+      newTurn = PLAYER_1;
+    }
+    setTurn(newTurn);
+  };
+
+  const onClickCallback = (squareID) => {
+    const newSquares = [...squares];
+    newSquares.forEach((row) => {
+      row.forEach((square) => {
+        if (square.id === squareID && square.value === '') {
+          square.value = turn;
+          ticDict[squareID] = turn;
+        }
+      });
+    });
+    setSquares(newSquares);
+    toggleTurn();
+
+    let someoneWon = checkForWinner();
+    if (someoneWon) {
+      setWinner(someoneWon);
+      squares.forEach((row) => {
+        row.map((square) => {
+          if (square.value == '') {
+            square.disabled = true;
+          }
+        });
+      });
+    }
+  };
+
   const checkForWinner = () => {
     // Complete in Wave 3
     // You will need to:
-    // 1. Go accross each row to see if
+    // 1. Go across each row to see if
     //    3 squares in the same row match
     //    i.e. same value
     // 2. Go down each column to see if
     //    3 squares in each column match
     // 3. Go across each diagonal to see if
     //    all three squares have the same value.
+    /* we need to update dict each click and then check 
+    
+    012, 345, 678 horizontal winners
+    036, 147, 258 vertical winners
+    048, 246 diagonal winners */
+
+    let winner = null;
```

Comment:
There is an interesting bug in this function!  The code fails the test that checks the bottom row of x's.  The test setup goes as follows: 6 - x, 1 - o, 8 - x, 2 - o, 7 - x, which results in 2 o's in the top row and 3 x's in the bottom row.  The if-else sequence to find the winner fails at this point, because line 94 is _true_!  For values 3, 4 and 5, ticDict is _undefined_, so they are equal!  As a result, winner is set to _undefined_.

Reply:
Oh my goodness. Thank you for figuring this one out--we were stumped!

Reply:
Haha, cool bug!!! Thanks for pointing it out :-)